pubsub/awssnssqs/awssnssqs.go

```diff
@@ -530,25 +531,50 @@ func (s *subscription) receiveMessages(ctx context.Context, max int64) ([]*drive
 
 // SendAcks implements driver.Subscription.SendAcks.
 func (s *subscription) SendAcks(ctx context.Context, ids []driver.AckID) error {
+	// SQS supports deleting at most 10 messages at a time:
+	// https://godoc.org/github.com/aws/aws-sdk-go/service/sqs#SQS.DeleteMessageBatch
+	// If len(ids) is larger than 10, we make up to maxAckConcurrency RPCs
+	// (across calls to SendAcks) in parallel.
+	const maxPerRPC = 10
+
 	g, ctx := errgroup.WithContext(ctx)
 	var ctxErr error
 
+	// sendBatch sends a single RPC deleting up to maxPerRPC messages.
+	sendBatch := func(req *sqs.DeleteMessageBatchInput) error {
+		defer func() { <-s.acksem }()
+		resp, err := s.client.DeleteMessageBatchWithContext(ctx, req)
+		if err != nil {
+			return err
+		}
+		if numFailed := len(resp.Failed); numFailed > 0 {
+			first := resp.Failed[0]
+			return awserr.New(aws.StringValue(first.Code), fmt.Sprintf("sqs.DeleteMessageBatch failed for %d message(s): %s", numFailed, aws.StringValue(first.Message)), nil)
+		}
+		return nil
+	}
+
+	req := &sqs.DeleteMessageBatchInput{QueueUrl: aws.String(s.qURL)}
 Loop:
-	for _, id := range ids {
-		select {
-		case <-ctx.Done():
-			ctxErr = ctx.Err()
-			break Loop
-		case s.acksem <- struct{}{}:
-			id := id
-			g.Go(func() error {
-				defer func() { <-s.acksem }()
-				_, err := s.client.DeleteMessageWithContext(ctx, &sqs.DeleteMessageInput{
-					QueueUrl:      &s.qURL,
-					ReceiptHandle: id.(*string),
+	for i, id := range ids {
+		req.Entries = append(req.Entries, &sqs.DeleteMessageBatchRequestEntry{
+			Id:            aws.String(strconv.Itoa(i)),
+			ReceiptHandle: id.(*string),
+		})
+		// Send if we've reached the limit, or if this is the last one.
+		if len(req.Entries) == maxPerRPC || i == len(ids)-1 {
+			select {
+			case <-ctx.Done():
+				ctxErr = ctx.Err()
+				break Loop
+			case s.acksem <- struct{}{}:
+				req := req
+				g.Go(func() error {
+					return sendBatch(req)
 				})
-				return err
-			})
+			}
+			// Start the next batch.
+			req = &sqs.DeleteMessageBatchInput{QueueUrl: aws.String(s.qURL)}
```

Comment:
or just `req.Entries = nil`.

Reply:
I don't think that works, because it leaves the `req` that we've handed to the background goroutine pointing to the same underlying struct, and we're modifying its fields.